### PR TITLE
Fix: clipping problem.

### PR DESCRIPTION
**Pull request by johanzhu**
In the pass, runtime dose not support spine animation with clip attachment. Now its work.

### CODE REVIEW

File: src/core/MeshGenerator.ts [hunk 1]
```diff
@@ -16,22 +16,21 @@ import { SpineRenderSetting } from '../types';
 
 export class MeshGenerator {
   static QUAD_TRIANGLES = [0, 1, 2, 2, 3, 0];
-  static VERTEX_SIZE = 2 + 2 + 4;
-  static VERTEX_STRIDE = 9;
+  static VERTEX_SIZE = 2 + 2 + 4; // position without z, uv, color
+  static VERTEX_STRIDE = 3 + 2 + 4; // position with z, uv, color
```
Comment (singlecoder) [inline src/core/MeshGenerator.ts:20]: static VERTEX_STRIDE = 9; // 3 2 4 position with z, uv, color

File: src/core/MeshGenerator.ts [hunk 2]
```diff
@@ -42,29 +41,47 @@ export class MeshGenerator {
...
-    const meshRenderer = this.entity.getComponent(MeshRenderer);
-    if (!meshRenderer) {
-      console.warn('You need add MeshRenderer component to entity first');
+    // Prepare buffer by using all attachment data but clippingAttachment
+    const { defaultSkin: { attachments } } = skeletonData;
+    let vertexCount: number = 0;
+    const attachmentsArray = [];
+    for (let i = 0, n = attachments.length; i < n; i++) {
+      for (let key in attachments[i]) {
+        attachmentsArray.push(attachments[i][key]);
```
Comment (singlecoder) [inline src/core/MeshGenerator.ts:63]: attachments[i] should put out of for(let key in attachments[i])

File: src/core/MeshGenerator.ts [hunk 3]
```diff
@@ -42,29 +41,47 @@ export class MeshGenerator {
...
+      }
+    }
+    for (let i = 0, n = attachmentsArray.length; i < n; i++) {
+      const attachment = attachmentsArray[i];
+      if (!attachment) {
+        continue;
+      } else if (attachment instanceof RegionAttachment) {
+        vertexCount += MeshGenerator.QUAD_TRIANGLES.length;
```
Comment (singlecoder) [inline src/core/MeshGenerator.ts:71]: MeshGenerator.QUAD_TRIANGLES.length cache before for is better? the same as mesh.triangles.length

File: src/core/MeshGenerator.ts [hunk 4]
```diff
@@ -42,29 +41,47 @@ export class MeshGenerator {
...
+      } else if (attachment instanceof MeshAttachment) {
+        let mesh = attachment;
+        vertexCount += mesh.triangles.length;
+      } else continue;
+    }
+    this.vertexCount = vertexCount;
+    this.prepareBufferData(this.vertexCount);
+    this.spineMesh.initialize(this.engine, this.vertexCount);
+    meshRenderer.mesh = this.spineMesh.mesh;
```
Comment (singlecoder) [inline src/core/MeshGenerator.ts:80]: spineMesh need cache

File: src/core/MeshGenerator.ts [hunk 5]
```diff
@@ -204,43 +222,39 @@ export class MeshGenerator {
...
+    if (this.needResize) {
+      // #1
+      this.spineMesh.vertexBuffer.resize(this.verticesWithZ.byteLength);
+      this.spineMesh.indexBuffer.resize(this.indices.byteLength);
+      // #2 https://github.com/oasis-engine/engine/issues/376
+      // this.spineMesh.changeBuffer(this.engine, this.vertexCount);
+      this.needResize = false;
+    }
+    this.spineMesh.vertexBuffer.setData(this.verticesWithZ);
+    this.spineMesh.indexBuffer.setData(this.indices);
```
Comment (singlecoder) [inline src/core/MeshGenerator.ts:247]: spineMesh should be cached

**Review by singlecoder — APPROVED** [review]
+1

File: src/core/MeshGenerator.ts [hunk 6]
```diff
@@ -42,29 +41,50 @@ export class MeshGenerator {
...
     }
 
-    const meshRenderer = this.entity.getComponent(MeshRenderer);
-    if (!meshRenderer) {
-      console.warn('You need add MeshRenderer component to entity first');
+    // Prepare buffer by using all attachment data but clippingAttachment
+    const { defaultSkin: { attachments } } = skeletonData;
+    let vertexCount: number = 0;
+    const QUAD_TRIANGLE_LENGTH = MeshGenerator.QUAD_TRIANGLES.length;
+    const attachmentsArray = [];
```
Comment (GuoLei1990) [inline src/core/MeshGenerator.ts:61]: GC

File: src/core/MeshGenerator.ts [hunk 7]
```diff
@@ -204,43 +224,39 @@ export class MeshGenerator {
...
+      // #2 https://github.com/oasis-engine/engine/issues/376
+      // spineMesh.changeBuffer(this.engine, this.vertexCount);
+      this.needResize = false;
+    }
+    spineMesh.vertexBuffer.setData(this.verticesWithZ);
+    spineMesh.indexBuffer.setData(this.indices);
   }
 
+
   addSubMesh(skeleton: Skeleton) {
```
Comment (GuoLei1990) [inline src/core/MeshGenerator.ts:253]: Delete

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**